fedimint-client/src/oplog.rs

```diff
@@ -0,0 +1,424 @@
+            extra_meta: serde_json::Value,
+        }
+
+        let meta = Meta {
+            foo: "bar".to_string(),
+            extra_meta: serde_json::to_value(()).unwrap(),
+        };
+
+        let op_log = OperationLogEntry {
+            operation_type: "test".to_string(),
+            meta: serde_json::to_value(meta.clone()).unwrap(),
+            outcome: None,
+        };
+
+        assert_eq!(op_log.meta::<Meta>(), meta);
+    }
+
+    #[tokio::test]
+    async fn test_operation_log_update() {
+        let op_id = OperationId([0x32; 32]);
+
+        let db = Database::new(MemDatabase::new(), Default::default());
+        let op_log = OperationLog::new(db.clone());
+
+        let mut dbtx = db.begin_transaction().await;
+        op_log
+            .add_operation_log_entry(&mut dbtx, op_id, "foo", "bar")
+            .await;
+        dbtx.commit_tx().await;
+
+        let op = op_log.get_operation(op_id).await.expect("op exists");
+        assert_eq!(op.outcome, None);
+
+        OperationLog::set_operation_outcome(&db, op_id, &"baz")
+            .await
+            .unwrap();
+
+        let op = op_log.get_operation(op_id).await.expect("op exists");
+        assert_eq!(op.outcome::<String>(), Some("baz".to_string()));
+
+        let update_stream_or_outcome =
+            op.outcome_or_updates::<String, _>(&db, op_id, futures::stream::empty);
+
+        assert!(matches!(
+            &update_stream_or_outcome,
+            UpdateStreamOrOutcome::Outcome(s) if s == "baz"
+        ));
+
+        let updates = update_stream_or_outcome
+            .into_stream()
+            .collect::<Vec<_>>()
+            .await;
+        assert_eq!(updates, vec!["baz"]);
+    }
+
+    #[tokio::test]
+    async fn test_pagination() {
+        let db = Database::new(MemDatabase::new(), Default::default());
+        let op_log = OperationLog::new(db.clone());
+
+        for operation_idx in 0u8..98 {
+            let mut dbtx = db.begin_transaction().await;
+            op_log
+                .add_operation_log_entry(
+                    &mut dbtx,
+                    OperationId([operation_idx; 32]),
+                    "foo",
+                    operation_idx,
+                )
+                .await;
+            dbtx.commit_tx().await;
+        }
+
+        fn assert_page_entries(
+            page: Vec<(ChronologicalOperationLogKey, OperationLogEntry)>,
+            page_idx: u8,
+        ) {
+            for (entry_idx, (_key, entry)) in page.into_iter().enumerate() {
+                let actual_meta = entry.meta::<u8>();
+                let expected_meta = 97 - (page_idx * 10 + entry_idx as u8);
+
+                assert_eq!(actual_meta, expected_meta);
+            }
+        }
+
+        let mut previous_last_element = None;
+        for page_idx in 0u8..9 {
+            let page = op_log.list_operations(10, previous_last_element).await;
+            previous_last_element = Some(page[9].0);
+            assert_eq!(page.len(), 10);
```

Comment:
assert length first, lookup later
```suggestion
            assert_eq!(page.len(), 10);
            previous_last_element = Some(page[9].0);
```

Reply:
Fixed, thx!